chat: enter key full functionality (fixes #9428)

emmanuelbaa opened the pull request:
fixes #9428 

<img width="1918" height="868" alt="make conf" src="https://github.com/user-attachments/assets/e0f926b8-10f2-4d15-995c-39632a8f3949" />

This commit introduces several key improvements to the chat feature to enhance user experience and fix bugs.

1.  **Enter-to-Send Functionality:**
    - Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.
    - Adds a `keydown` event listener to the chat textarea to handle this logic.
    - Replicates the existing form validation from the send button to prevent empty messages or multiple submissions.

2.  **Immediate UI Feedback (Optimistic Update):**
    - The chat UI is now updated optimistically. The user's message is added to the conversation list and the input field is cleared as soon as the message is submitted, rather than waiting for the server response. This makes the chat feel more responsive.

3.  **Sidebar Conversation History Fix:**
    - Fixes a bug where new conversations were not appearing in the sidebar history until after the first response was received.
    - A signal is now sent to the `ChatService` as soon as a message is sent in a new conversation, ensuring the sidebar updates immediately.
    - Fixes a subsequent bug where this signal was being sent twice.

4.  **UI Rendering Fix:**
    - Restores the `role: 'user'` property to the conversation object when performing an optimistic update, preventing potential UI rendering issues for user messages.

google-labs-jules[bot] committed:
feat(chat): Send message on Enter key press

Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.

- Adds a `keydown` event listener to the chat textarea.
- The event handler checks for the "Enter" key press without the "Shift" key.
- Replicates the existing form validation from the send button to ensure empty messages or multiple submissions while processing are not allowed.

google-labs-jules[bot] committed:
feat(chat): Send message on Enter key press

Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.

- Adds a `keydown` event listener to the chat textarea.
- The event handler checks for the "Enter" key press without the "Shift" key.
- Replicates the existing form validation from the send button to ensure empty messages or multiple submissions while processing are not allowed.

google-labs-jules[bot] committed:
feat(chat): Send message on Enter and provide immediate feedback

Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.

- Adds a `keydown` event listener to the chat textarea.
- The event handler checks for the "Enter" key press without the "Shift" key and sends the message.
- Replicates the existing form validation from the send button to ensure empty messages or multiple submissions while processing are not allowed.

Additionally, this change improves the user experience by providing immediate feedback when a message is sent.

- The chat UI is now updated optimistically, adding the user's message to the conversation list and clearing the input field as soon as the message is sent, rather than waiting for the server response. This makes the chat feel more responsive.

google-labs-jules[bot] committed:
feat(chat): Send message on Enter and provide immediate feedback

Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.

- Adds a `keydown` event listener to the chat textarea.
- The event handler checks for the "Enter" key press without the "Shift" key and sends the message.
- Replicates the existing form validation from the send button to ensure empty messages or multiple submissions while processing are not allowed.

Additionally, this change improves the user experience by providing immediate feedback when a message is sent.

- The chat UI is now updated optimistically, adding the user's message to the conversation list and clearing the input field as soon as the message is sent, rather than waiting for the server response. This makes the chat feel more responsive.

emmanuelbaa committed:
Merge branch '9428-chat-enterKeyFunction' of https://github.com/open-learning-exchange/planet into 9428-chat-enterKeyFunction

google-labs-jules[bot] committed:
feat(chat): Implement Enter-to-send, immediate feedback, and sidebar fix

This commit introduces several key improvements to the chat feature to enhance user experience and fix bugs.

1.  **Enter-to-Send Functionality:**
    - Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.
    - Adds a `keydown` event listener to the chat textarea to handle this logic.
    - Replicates the existing form validation from the send button to prevent empty messages or multiple submissions.

2.  **Immediate UI Feedback (Optimistic Update):**
    - The chat UI is now updated optimistically. The user's message is added to the conversation list and the input field is cleared as soon as the message is submitted, rather than waiting for the server response. This makes the chat feel more responsive.

3.  **Sidebar Conversation History Fix:**
    - Fixes a bug where new conversations were not appearing in the sidebar history until after the first response was received.
    - A signal is now sent to the `ChatService` as soon as a message is sent in a new conversation, ensuring the sidebar updates immediately.
    - Fixes a subsequent bug where this signal was being sent twice.

4.  **UI Rendering Fix:**
    - Restores the `role: 'user'` property to the conversation object when performing an optimistic update, preventing potential UI rendering issues for user messages.

google-labs-jules[bot] committed:
feat(chat): Implement Enter-to-send, immediate feedback, and sidebar fix

This commit introduces several key improvements to the chat feature to enhance user experience and fix bugs.

1.  **Enter-to-Send Functionality:**
    - Implements the standard chat behavior where the "Enter" key sends the message and "Shift + Enter" creates a new line.
    - Adds a `keydown` event listener to the chat textarea to handle this logic.
    - Replicates the existing form validation from the send button to prevent empty messages or multiple submissions.

2.  **Immediate UI Feedback (Optimistic Update):**
    - The chat UI is now updated optimistically. The user's message is added to the conversation list and the input field is cleared as soon as the message is submitted, rather than waiting for the server response. This makes the chat feel more responsive.

3.  **Sidebar Conversation History Fix:**
    - Fixes a bug where new conversations were not appearing in the sidebar history until after the first response was received.
    - A signal is now sent to the `ChatService` as soon as a message is sent in a new conversation, ensuring the sidebar updates immediately.
    - Fixes a subsequent bug where this signal was being sent twice.

4.  **UI Rendering Fix:**
    - Restores the `role: 'user'` property to the conversation object when performing an optimistic update, preventing potential UI rendering issues for user messages.

Mutugiii reviewed (changes requested):
This brings about some regressions in the chat:
- On submit, we see the last created chat instead of the question asked
- Newly created chats are no longer showing up in the sidebar
- Enter submission disregards the chat disabled state

emmanuelbaa commented:
> This brings about some regressions in the chat:
> 
> * On submit, we see the last created chat instead of the question asked
> * Newly created chats are no longer showing up in the sidebar
> * Enter submission disregards the chat disabled state

great reviews, i will work on that

emmanuelbaa committed:
Merge branch 'feature/chat-enterKeyFunction' into 9428-chat-enterKeyFunction

emmanuelbaa committed:
fix(chat): Prevent Enter submission when chat is disabled

Modify the onKeyDown method in ChatWindowComponent to check the
'disabled' state of the chat before allowing submission via the
Enter key. This prevents unintended submissions when the chat
input is visually or programmatically disabled.

emmanuelbaa committed:
fix(chat): Prevent Enter submission when chat is disabled

Modify the onKeyDown method in ChatWindowComponent to check the
'disabled' state of the chat before allowing submission via the
Enter key. This prevents unintended submissions when the chat
input is visually or programmatically disabled.

emmanuelbaa committed:
fix(chat): Resolve chat display and sidebar update regressions

Refactor submitPrompt() and postSubmit() in ChatWindowComponent to:
- Prevent premature clearing of conversation messages for new chats.
- Ensure the sidebar is notified correctly when a new chat is created
  and saved.

This addresses regressions where the last asked question was not
displaying on submit, and the last chat was not showing in the
sidebar.

emmanuelbaa committed:
fix(chat): Resolve chat display and sidebar update regressions

Refactor submitPrompt() and postSubmit() in ChatWindowComponent to:
- Prevent premature clearing of conversation messages for new chats.
- Ensure the sidebar is notified correctly when a new chat is created
  and saved.

This addresses regressions where the last asked question was not
displaying on submit, and the last chat was not showing in the
sidebar.

emmanuelbaa committed:
Update chat-window.component.ts

emmanuelbaa committed:
fix(chat): Ensure Enter submission respects chat disabled state

This commit re-applies the fix for the issue where pressing Enter
would submit a chat message even when the chat input was in a
disabled state. The onKeyDown method in ChatWindowComponent now
correctly checks the 'disabled' property before allowing submission.

emmanuelbaa committed:
fix(chat): Ensure Enter submission respects chat disabled state

This commit re-applies the fix for the issue where pressing Enter
would submit a chat message even when the chat input was in a
disabled state. The onKeyDown method in ChatWindowComponent now
correctly checks the 'disabled' property before allowing submission.

emmanuelbaa committed:
fix(chat): Resolve chat display and sidebar update regressions (re-attempt)

Re-addresses the regressions where the last asked question was not
displaying on submit and the last chat was not showing in the
sidebar.

- Modified subscribeToSelectedConversation() to filter out
  unnecessary re-fetches from the database after a local update,
  preventing overwrites of the current conversation.
- Corrected TypeScript error TS2339 by casting conversationId to
  any within the filter operator to allow access to the _id
  property.
- Modified submitPrompt() to correctly handle new conversation
  initiation and postSubmit() to conditionally send new chat
  signals, preventing double conversation resets and ensuring
  correct chat display and sidebar updates.

emmanuelbaa committed:
fix(chat): Resolve chat display and sidebar update regressions (re-attempt)

Re-addresses the regressions where the last asked question was not
displaying on submit and the last chat was not showing in the
sidebar.

- Modified subscribeToSelectedConversation() to filter out
  unnecessary re-fetches from the database after a local update,
  preventing overwrites of the current conversation.
- Corrected TypeScript error TS2339 by casting conversationId to
  any within the filter operator to allow access to the _id
  property.
- Modified submitPrompt() to correctly handle new conversation
  initiation and postSubmit() to conditionally send new chat
  signals, preventing double conversation resets and ensuring
  correct chat display and sidebar updates.

emmanuelbaa committed:
Merge branch '9428-chat-enterKeyFunction' of https://github.com/open-learning-exchange/planet into 9428-chat-enterKeyFunction

emmanuelbaa committed:
fix(chat): resolve merge conflicts in ChatWindowComponent

Saby-Bishops reviewed (changes requested):
There still seems to be some regressions where Enter triggers submission even when the action is disabled, and where new chats don’t always update the sidebar as expected.

emmanuelbaa committed:
fix(validator): annotate currentTime map params as number to resolve TS2345

emmanuelbaa committed:
fix(chat): prevent Enter from bypassing send-disabled state; notify sidebar immediately when creating new conversation

emmanuelbaa committed:
fix(chat): avoid premature sidebar refresh that replaced in-progress conversation; defer new-chat signal until save completes
